wasm/wasm_source/src/vp_user.rs

```diff
@@ -158,35 +136,200 @@ fn validate_tx(
                     is_vp_whitelisted(ctx, &vp_hash)?
                 }
             }
-            KeyType::Masp => true,
+            KeyType::Masp | KeyType::Ibc => true,
             KeyType::Unknown => {
-                if key.segments.get(0) == Some(&addr.to_db_key()) {
-                    // Unknown changes to this address space require a valid
-                    // signature
-                    *valid_sig
-                } else {
-                    // Unknown changes anywhere else are permitted
-                    true
-                }
+                // Unknown changes require a valid signature
+                *valid_sig
             }
         };
         if !is_valid {
-            debug_log!("key {} modification failed vp", key);
+            log_string(format!("key {} modification failed vp_user", key));
             return reject();
         }
     }
 
     accept()
 }
 
+fn validate_pos_changes(
+    ctx: &Ctx,
+    owner: &Address,
+    key: &storage::Key,
+    valid_sig: &impl Deref<Target = bool>,
+) -> VpResult {
+    use proof_of_stake::storage;
+
+    // Bond or unbond
+    let is_valid_bond_or_unbond_change = || {
+        let bond_id = storage::is_bond_key(key)
+            .map(|(bond_id, _)| bond_id)
+            .or_else(|| storage::is_bond_epoched_meta_key(key))
+            .or_else(|| {
+                storage::is_unbond_key(key).map(|(bond_id, _, _)| bond_id)
+            });
+        if let Some(bond_id) = bond_id {
+            // Bonds and unbonds changes for this address must be signed
+            return &bond_id.source != owner || **valid_sig;
```

Comment:
Should the `&bond_id.source != owner` still be here now? How do we enforce that a delegation is signed / truly submitted by the delegator?

Reply:
I believe that’s what this does - if the source is eq to the owner then the sig is required

Reply:
ah yes ok, forgot that this is now the VP for normal users and validators. So in the delegator's VP, the sig will be required, but in the validator's VP it won't be, as expected